### PR TITLE
feat(frontend/backend): quiz functionality bb-89

### PR DESCRIPTION
**Pull request by Mikoyzskie**
-  Get all questions with answers from Backend
-  Group questions by categories
-  Display questions group by group

### CODE REVIEW

File: packages/shared/tsconfig.json [hunk 1]
```diff
@@ -1,6 +1,10 @@
 {
 	"extends": "../../tsconfig.json",
-	"include": ["**/*.ts", "**/*.js"],
+	"include": [
+		"**/*.ts",
+		"**/*.js",
+		"../../apps/backend/src/modules/quiz/libs/types/quiz-questions-get-all-response.type.ts"
```
Comment (ihorLysak) [inline packages/shared/tsconfig.json:6]: This shouldn't be here

File: packages/shared/src/modules/quiz/libs/types/quiz-answers-dto.type.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+type QuizAnswersDto = {
```
Comment (ihorLysak) [inline packages/shared/src/modules/quiz/libs/types/quiz-answers-dto.type.ts:1]: Looks like a Dto for a single entity, maybe QuizAnswerDto would be better?

File: packages/shared/src/modules/quiz/libs/types/quiz-questions-dto.type.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { type QuizAnswersDto } from "./quiz-answers-dto.type.js";
+
+type QuizQuestionsDto = {
```
Comment (ihorLysak) [inline packages/shared/src/modules/quiz/libs/types/quiz-questions-dto.type.ts:3]: same as with QuizAnswersDto

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/styles.module.css [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	align-items: center;
+	justify-content: center;
+}
+
+.btn button,
+.btn-secondary button {
+	padding: 10px 35px;
+}
+
+.btn-secondary button {
```
Comment (ihorLysak) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/styles.module.css:39]: maybe move this to Button container?

File: apps/frontend/src/pages/quiz/libs/components/progress-bar/progress-bar.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import React from "react";
```
Comment (ihorLysak) [inline apps/frontend/src/pages/quiz/libs/components/progress-bar/progress-bar.tsx:1]: We can use React without importing it

File: apps/backend/src/modules/quiz/libs/types/quiz-answers-dto.type.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+type QuizAnswersDto = {
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/libs/types/quiz-answers-dto.type.ts:1]: I believe this should be in shared

Comment (fshabanov) [inline apps/backend/src/modules/quiz/libs/types/quiz-answers-dto.type.ts:1]:
```suggestion
type QuizAnswerDto = {
```

File: apps/backend/src/modules/quiz/libs/types/quiz-questions-dto.type.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { type QuizAnswersDto } from "./quiz-answers-dto.type.js";
+
+type QuizQuestionsDto = {
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/libs/types/quiz-questions-dto.type.ts:3]: Should be in shared

File: apps/backend/src/modules/quiz/quiz-questions.entity.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
+import { type QuizAnswersDto } from "./libs/types/types.js";
+
+class QuizQuestionsEntity {
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/quiz-questions.entity.ts:3]:
```suggestion
class QuizQuestionsEntity implements Entity {
```

File: apps/backend/src/modules/quiz/quiz.repository.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import { RelationName } from "~/libs/enums/enums.js";
+import { QuizQuestionsEntity } from "~/modules/quiz/quiz-questions.entity.js";
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/quiz.repository.ts:2]:
```suggestion
import { QuizQuestionsEntity } from "./quiz-questions.entity.js";
```

File: apps/backend/src/modules/quiz/quiz.repository.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+import { type QuizQuestionsModel } from "./quiz-questions.model.js";
+
+class QuizQuestionsRepository {
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/quiz.repository.ts:6]:
```suggestion
class QuizQuestionsRepository implements Repository {
```

File: apps/backend/src/modules/quiz/quiz.service.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import { type QuizQuestionsRepository } from "~/modules/quiz/quiz.repository.js";
+
+import { type QuizQuestionsDto } from "./libs/types/types.js";
+
+class QuizQuestionsService {
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/quiz.service.ts:5]:
```suggestion
class QuizQuestionsService implements Service {
```

File: apps/frontend/src/libs/components/quiz-question/quiz-question.tsx [hunk 12]
```diff
@@ -9,18 +9,24 @@ import styles from "./styles.module.css";
 
 type Properties<T extends FieldValues> = {
 	control: Control<T, null>;
+	label: string;
+	name: string;
+	options?: { label: string; value: string }[];
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/quiz-question/quiz-question.tsx:14]: Create a separate type for that

File: apps/frontend/src/modules/quiz/slices/quiz.slice.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+type State = {
+	dataStatus: ValueOf<typeof DataStatus>;
+	questions: null | QuizQuestionsDto[];
+};
+
+const initialState: State = {
+	dataStatus: DataStatus.IDLE,
+	questions: null,
+};
```
Comment (fshabanov) [inline apps/frontend/src/modules/quiz/slices/quiz.slice.ts:17]:
```suggestion
type State = {
	dataStatus: ValueOf<typeof DataStatus>;
	questions: QuizQuestionsDto[];
};

const initialState: State = {
	dataStatus: DataStatus.IDLE,
	questions: [],
};
```

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 14]
```diff
@@ -1,19 +1,136 @@
-import { QuizQuestion } from "~/libs/components/components.js";
-import { useAppForm } from "~/libs/hooks/hooks.js";
+import { Button, Loader, QuizQuestion } from "~/libs/components/components.js";
+import { DataStatus } from "~/libs/enums/data-status.enum.js";
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:2]:
```suggestion
import { DataStatus } from "~/libs/enums/enums.js";
```

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 15]
```diff
@@ -1,19 +1,136 @@
...
+	useAppSelector,
+	useCallback,
+	useEffect,
+	useState,
+} from "~/libs/hooks/hooks.js";
+import {
+	actions as quizActions,
+	type QuizQuestionsDto,
+} from "~/modules/quiz/quiz.js";
+import { ProgressBar } from "~/pages/quiz/libs/components/components.js";
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:15]: Move to ~/libs/components/components

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 16]
```diff
@@ -1,19 +1,136 @@
...
+					groupedByCategory[categoryId] = [];
+				}
+
+				groupedByCategory[categoryId].push(question);
+			}
+
+			const result = Object.values(groupedByCategory);
+			setCategories(result);
+		}
+	}, [questions]);
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:64]: Too much logic for frontend, should be done in backend

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/styles.module.css [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+.form-footer {
+	display: flex;
+	gap: 15px;
+	align-items: center;
+	justify-content: center;
+}
+
+.btn button,
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/styles.module.css:34]: B10

File: packages/shared/src/modules/quiz/libs/enums/quiz-api-path.enum.ts [hunk 18]
```diff
@@ -0,0 +1,6 @@
+const QuizApiPath = {
+	QUIZ_ANSWER: "/answers",
+	QUIZ_QUESTION: "/questions",
```
Comment (fshabanov) [inline packages/shared/src/modules/quiz/libs/enums/quiz-api-path.enum.ts:3]:
```suggestion
	ANSWERS: "/answers",
	QUESTIONS: "/questions",
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/modules/quiz/quiz-api.ts [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+	http: HTTP;
+	storage: Storage;
+};
+
+class QuizApi extends BaseHTTPApi {
+	public constructor({ baseUrl, http, storage }: Constructor) {
+		super({ baseUrl, http, path: APIPath.QUIZ, storage });
+	}
+
+	public async getQuestions(): Promise<QuizQuestionsGetAllReponseDto> {
```
Comment (ihorLysak) [inline apps/frontend/src/modules/quiz/quiz-api.ts:20]: maybe getAllQuestions? Feels more descriptive to me

File: apps/backend/src/modules/quiz/quiz.service.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+	public async findAll(): Promise<QuizQuestionsGetAllReponseDto> {
+		const items = await this.quizQuestionsRepository.findAll();
+
+		return {
+			items: items.map((item) => item.toObject()),
+		};
+	}
+
+	update(): Promise<unknown> {
```
Comment (ihorLysak) [inline apps/backend/src/modules/quiz/quiz.service.ts:43]: Why not?

File: apps/backend/src/libs/modules/server-application/libs/constants/constants.ts [hunk 21]
```diff
@@ -7,6 +8,7 @@ const WHITE_ROUTES: string[] = [
 	APIPath.AUTH,
 	`${APIPath.AUTH}${AuthApiPath.SIGN_IN}`,
 	`${APIPath.AUTH}${AuthApiPath.SIGN_UP}`,
+	`${APIPath.QUIZ}${QuizApiPath.QUESTION}`,
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/libs/constants/constants.ts:11]: It shouldn't be a white route

File: apps/backend/src/modules/quiz/quiz-questions.entity.ts [hunk 22]
```diff
@@ -0,0 +1,116 @@
+import { type Entity } from "~/libs/types/entity.type.js";
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/quiz-questions.entity.ts:1]:
```suggestion
import { type Entity } from "~/libs/types/types.js";
```

File: apps/backend/src/modules/quiz/quiz.repository.ts [hunk 23]
```diff
@@ -0,0 +1,84 @@
+import { RelationName } from "~/libs/enums/enums.js";
+import { type Repository } from "~/libs/types/repository.type.js";
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/quiz.repository.ts:2]:
```suggestion
import { type Repository } from "~/libs/types/types.js";
```

File: apps/backend/src/modules/quiz/quiz.repository.ts [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+				})
+			: null;
+	}
+
+	public async findAll(): Promise<QuizQuestionsEntity[]> {
+		const questions = await this.quizQuestionsModel
+			.query()
+			.withGraphFetched(RelationName.QUIZ_ANSWERS);
+
+		return questions.map((question) =>
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/quiz.repository.ts:67]: B27

File: apps/backend/src/modules/quiz/quiz.service.ts [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+import { QuizQuestionsEntity } from "./quiz-questions.entity.js";
+
+class QuizQuestionsService implements Service {
+	private quizQuestionsRepository: QuizQuestionsRepository;
+
+	public constructor(quizQuestionsRepository: QuizQuestionsRepository) {
+		this.quizQuestionsRepository = quizQuestionsRepository;
+	}
+	public async create(payload: QuizQuestionDto): Promise<QuizQuestionDto> {
+		const user = await this.quizQuestionsRepository.create(
```
Comment (fshabanov) [inline apps/backend/src/modules/quiz/quiz.service.ts:17]: Why does creating a quiz question return a user?

File: apps/backend/src/modules/quiz/quiz.service.ts (hunk 20 above)
Comment (fshabanov) [reply apps/backend/src/modules/quiz/quiz.service.ts:43]: Should be implemented

File: apps/frontend/src/libs/components/input/input.tsx [hunk 26]
```diff
@@ -11,14 +11,19 @@ import {
 import { Button } from "../button/button.js";
 import styles from "./styles.module.css";
 
+type options = {
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:14]: Should be in a separate file, capitalized

File: apps/frontend/src/libs/components/progress-bar/progress-bar.tsx [hunk 27]
```diff
@@ -0,0 +1,54 @@
+import CheckIcon from "~/assets/img/check-icon.svg?react";
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/progress-bar/progress-bar.tsx:1]: We have a separate Icon component

File: apps/frontend/src/libs/components/progress-bar/progress-bar.tsx [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+	steps: number;
+};
+
+const LAST_INDEX_OFFSET = 1;
+
+const ProgressBar: React.FC<Properties> = ({
+	currentStep,
+	steps,
+}: Properties) => {
+	const elementsArray = Array.from({ length: steps });
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/progress-bar/progress-bar.tsx:17]: B17

File: apps/frontend/src/libs/components/quiz-question/quiz-question.tsx [hunk 29]
```diff
@@ -7,20 +7,31 @@ import {
 import { Input } from "../input/input.js";
 import styles from "./styles.module.css";
 
+type options = {
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/quiz-question/quiz-question.tsx:10]: Should be in a separate file. Same type for input as well

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 30]
```diff
@@ -1,19 +1,143 @@
-import { QuizQuestion } from "~/libs/components/components.js";
-import { useAppForm } from "~/libs/hooks/hooks.js";
+import {
+	Button,
+	Loader,
+	ProgressBar,
+	QuizQuestion,
+} from "~/libs/components/components.js";
+import { DataStatus } from "~/libs/enums/data-status.enum.js";
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:7]:
```suggestion
import { DataStatus } from "~/libs/enums/enums.js";
```

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 31]
```diff
@@ -1,19 +1,143 @@
...
+					groupedByCategory[categoryId] = [];
+				}
+
+				groupedByCategory[categoryId].push(question);
+			}
+
+			const result = Object.values(groupedByCategory);
+			setCategories(result);
+		}
+	}, [dataStatus, questions]);
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:68]: Too much logic for Frontend, should be in Backend

File: packages/shared/src/modules/quiz/libs/enums/quiz-api-path.enum.ts [hunk 32]
```diff
@@ -1,5 +1,6 @@
 const QuizApiPath = {
 	ANSWER: "/answer",
+	QUESTION: "/question",
```
Comment (fshabanov) [inline packages/shared/src/modules/quiz/libs/enums/quiz-api-path.enum.ts:3]:
```suggestion
	QUESTIONS: "/questions",
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/input/input.tsx [hunk 33]
```diff
@@ -6,6 +6,7 @@ import {
 	type FieldPath,
 	type FieldValues,
 	type IconName,
+	type OPTIONS,
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/input/input.tsx:9]:
```suggestion
	type Options,
```

File: apps/frontend/src/libs/components/quiz-question/quiz-question.tsx [hunk 34]
```diff
@@ -2,25 +2,32 @@ import {
 	type Control,
 	type FieldPath,
 	type FieldValues,
+	type OPTIONS,
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/quiz-question/quiz-question.tsx:5]:
```suggestion
	type Options,
```

File: apps/frontend/src/modules/quiz/slices/quiz.slice.ts [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+
+import { DataStatus } from "~/libs/enums/enums.js";
+import { type ValueOf } from "~/libs/types/types.js";
+import { type QuizQuestionDto } from "~/modules/quiz/quiz.js";
+
+import { getAllQuestions } from "./actions.js";
+
+type State = {
+	dataStatus: ValueOf<typeof DataStatus>;
+	questions: { items: QuizQuestionDto[][] };
```
Comment (fshabanov) [inline apps/frontend/src/modules/quiz/slices/quiz.slice.ts:11]:
```suggestion
	questionsByCategories: QuizQuestionDto[][]
```

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 36]
```diff
@@ -1,19 +1,111 @@
...
+
+	useEffect(() => {
+		void dispatch(quizActions.getAllQuestions());
+	}, [dispatch]);
+
+	useEffect(() => {
+		setIsLast(currentStep === questions.length - ONE_STEP_OFFSET);
+	}, [currentStep, questions]);
+
+	const handleBackStep = useCallback(() => {
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:52]:
```suggestion
	const handlePreviousStep = useCallback(() => {
```

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 37]
```diff
@@ -1,19 +1,111 @@
...
+
+	const isLoading = dataStatus === DataStatus.PENDING;
+
 	return (
-		<form>
-			<QuizQuestion control={control} />
-		</form>
+		<div className={styles["quiz-container"]}>
+			<form className={styles["questions-form"]}>
+				<ProgressBar currentStep={currentStep} steps={questions.length} />
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:66]:
```suggestion
				<ProgressBar currentStep={currentStep} numberOfSteps={questions.length} />
```

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 38]
```diff
@@ -1,19 +1,111 @@
...
+						questions[currentStep]?.map((question) => (
+							<QuizQuestion
+								control={control}
+								key={question.id}
+								label={`${question.id.toString()}. ${question.label}`}
+								name={question.label}
+								options={question.answers.map(({ label, value }) => ({
+									label,
+									value: value.toString(),
+								}))}
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:83]: Move this to a separate variable

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/styles.module.css (hunk 17 above)
Comment (fshabanov) [reply apps/frontend/src/pages/quiz/libs/components/quiz-form/styles.module.css:34]: Not resolved

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/types/input-options.type.ts [hunk 39]
```diff
@@ -0,0 +1,6 @@
+type Options = {
```
Comment (fshabanov) [inline apps/frontend/src/libs/types/input-options.type.ts:1]:
```suggestion
type RadioInputOption = {
```
Change the file name as well

File: apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx [hunk 40]
```diff
@@ -1,19 +1,115 @@
...
+
+const ONE_STEP_OFFSET = 1;
+const INITIAL_STEP = 0;
+
+const QuizForm: React.FC<Properties> = ({ onNext }: Properties) => {
+	const dispatch = useAppDispatch();
+	const [isLast, setIsLast] = useState<boolean>(false);
+	const [currentStep, setCurrentStep] = useState<number>(INITIAL_STEP);
 	const { control } = useAppForm<FormValues>({
 		defaultValues: { value: "" },
```
Comment (fshabanov) [inline apps/frontend/src/pages/quiz/libs/components/quiz-form/quiz-form.tsx:36]: It doesn't look like a valid default value, we shouldn't have mock values

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by fshabanov — APPROVED** [review]

**Review by fshabanov — APPROVED** [review]

**Pull request merged**